### DISCUSSION TITLE
Fron-end Development

### DISCUSSION THREAD

**Discussion by dingdingwall** [community/community#77947] [Programming Help] [answered]
### Body

Hi Guys, what is the roadmap for front-end development?  I know the basics of html,css, javascript, react-js, tailwind, and bootstrap. What should I learn next? 

### Guidelines

- [X] I have read and understood this category's [guidelines](https://github.com/orgs/community/discussions/51384) before making this post.

**Comment by ghost** [comment, marked as answer]
Hi @dingdingwall  hope your doing well.

You're almost there becoming a frontend developer.

I think you need to deliver deep into everything you learnt so far

You need to do the following:-
- Start building web pages or web apps , implementing the stuff you learnt
- Get a little intro to backend development
- Learn more about JS and it's asynchronous nature
- Learn about UI/UX if you didn't 
- Learn about JS Testing tools
- Also learn Typescript
- Build responsive websites
- And finally contribute to various open source projects in your field to level up your exp

I am not a front end web dev yet but when I was interested in it at first I watched the following video which helped me.

Video : https://youtu.be/9He4UBLyk8Y?si=uaY22hEGnAeZpU-H

Anyways hope it helps you.

And all the best in your Frontend Web Development Journey.

Thank you

**Reply by dingdingwall** [reply]
@Yuheshpandian Man, I still have a lot to learn pew. Thank you sm.

**Reply by ghost** [reply]
Thank you. I'm sure you would succeed the heights of front end Development. All the best in your learning journey.

**Comment by atarverd** [comment]
Hey @dingdingwall 
Hope this will help you : https://roadmap.sh/frontend
Good luck!

**Reply by dingdingwall** [reply]
Thank you

**Comment by gitfullstacker** [comment]
Hey there! Sounds like you've got a solid foundation in front-end development with HTML, CSS, JavaScript, React.js, Tailwind, and Bootstrap. Here are some suggestions for your next steps:

1. **Advanced JavaScript Concepts:** Dive deeper into JavaScript with topics like asynchronous programming, ES6+ features, functional programming concepts, and design patterns.

2. **TypeScript:** Consider learning TypeScript, a superset of JavaScript that adds static typing, improving code quality and maintainability.

3. **State Management:** Explore advanced state management libraries for React.js like Redux or Context API to handle complex application states efficiently.

4. **Performance Optimization:** Learn about performance optimization techniques, such as code splitting, lazy loading, and optimizing rendering for faster web applications.

5. **Responsive Design:** Deepen your understanding of responsive web design principles and techniques to ensure your websites look great on various devices.

6. **GraphQL:** Explore GraphQL as an alternative to REST APIs for fetching and manipulating data, providing more flexibility and efficiency in data fetching.

7. **Testing:** Familiarize yourself with testing frameworks like Jest for JavaScript/React testing and understand different testing methodologies (unit testing, integration testing).

8. **Progressive Web Apps (PWAs):** Learn about PWAs to build web applications that offer a native app-like experience, including offline functionality and push notifications.

9. **Build Tools and Task Runners:** Gain proficiency in build tools like Webpack and task runners like Gulp to automate tasks and optimize your development workflow.

10. **Version Control Systems:** If you haven't already, dive into Git for version control, branching strategies, and collaborative development practices.

Remember, the best path forward depends on your specific interests and the type of projects you aim to work on. Don't forget to practice and build projects to reinforce your learning!

**Comment by dzenebieri** [comment]
Next.js?

**Reply by ghost** [reply]
Yeah it is actually a good suggestion, as it simplifies react js and is efficient.

**Comment by B-Mani-kandan** [comment]
I have a doubt which is best Javascript library like React, Angular, Vue

**Reply by ghost** [reply]
It depends on the organization that you are working for. Each organization have different js Framework/Library requirement.

React is most popular and many Web devs use it currently so if you stuck with React js there is a large community to help you..

Angular is also a good choice as there is a large count of companies using it..

Vue had said to be having a significant faster community and users growth..

React is Developed by Facebook and Angular is developed by Google...


But if you ask me, I would suggest to opt React js as it is currently used by many organizations as many features like hooks, reusable code components etc.

**Comment by ItsJohnny30** [comment]
Vue / Nuxt.js?

**Comment by AbhisekhNayek** [comment]
Great start! After mastering HTML, CSS, JavaScript, React.js, Tailwind, and Bootstrap, consider diving deeper into advanced React concepts, state management with tools like Redux, and exploring Next.js for server-side rendering. Enhance your knowledge of responsive design, learn testing frameworks like Jest, and familiarize yourself with GraphQL for efficient data fetching. Additionally, delve into Webpack for bundling and optimizing your projects. Stay updated on emerging technologies and design principles while consistently building real-world projects to solidify your skills and stay competitive in front-end development.

1. Advanced React Concepts
   Dive into advanced React topics like Hooks, Context API, and React Router.

2. State Management:
   Explore state management libraries like Redux to efficiently manage the state of your applications.

3. Next.js
   Learn Next.js for server-side rendering and building more complex React applications with improved performance.

4. Responsive Design
   Enhance your skills in responsive design to create seamless user experiences across various devices.

5. Testing
   Familiarize yourself with testing frameworks like Jest for unit and integration testing of your React applications.

6. GraphQL
   Understand GraphQL for more efficient and flexible data fetching in your front-end projects.

7. Webpack
   Learn Webpack for bundling and optimizing your project's assets, improving performance and load times.

8. Stay Updated
   Keep yourself informed about emerging front-end technologies, updates, and best practices in the industry.

9. Design Principles
   Explore design principles and methodologies to improve the aesthetics and user experience of your applications.

10. Build Real Projects
   Apply your knowledge by consistently working on real-world projects to strengthen your skills and create a portfolio.

**Reply by ghost** [reply]
great advice!
